### PR TITLE
feat(date-picker): adds min/max date boundries for datepicker

### LINKED ISSUE #1984: [UX] Datepicker Min/Max Dates
**Select one ...**  (check one with "x")
```
[ ] bug
[ ] feature request
[x] enhancement
```

Add support for the following features to the Clarity Datepicker Component:
1. Min Dates
2. Max Dates

How it works on Chrome:
https://jsfiddle.net/rqjboweo/

### PR DESCRIPTION
**Pull request by hippee-lee**
- adds inputs for native HTML5 attributes min and max
- adds methods to DateIOService that handle user input for min/max: valid string, undefined, null and ''
- updates ClrDateInputSpec to cover min/max inputs
- updates DateIOServiceSpecs to cover min/maxUpdate methods
- closes #1984
Signed-off-by: Matt Hippely <mhippely@vmware.com>

Signed-off-by: Marcu Alexandru <alexandru.marcu@bytex.ro>

## PR Checklist

Please check if your PR fulfills the following requirements:

* [ ] Tests for the changes have been added (for bug fixes / features)
* [ ] Docs have been added / updated (for bug fixes / features)
* [ ] If applicable, have a visual design approval

## PR Type

What kind of change does this PR introduce?

* [ ] Bugfix
* [x] Feature
* [ ] Code style update (formatting, local variables)
* [ ] Refactoring (no functional changes, no api changes)
* [ ] Build related changes
* [ ] CI related changes
* [ ] Documentation content changes
* [ ] clarity.design website / infrastructure changes
* [ ] Other... Please describe:

## What is the current behavior?
There is no way to restrict date selection to a range. 

Issue Number: #1984

## What is the new behavior?
ClrDatePicker can use the native min/max inputs to restrict the date selection to a range. 

## Does this PR introduce a breaking change?

* [ ] Yes
* [x] No

### CODE REVIEW

File: src/clr-angular/forms/datepicker/model/calendar-view.model.ts [hunk 1]
```diff
@@ -80,12 +82,21 @@ export class CalendarViewModel {
     this.initializeFocusableDay();
   }
 
+  private isDateDisabled(date: DayModel) {
```
Comment (bdryanovski) [inline src/clr-angular/forms/datepicker/model/calendar-view.model.ts]: we changed the `isDisabled` to `isExcluded` we could continue this and for:

```ts
isDateDisabled() => isDateExcluded()
disabledDates => excludedDates
```
And not confuse them with `disabled` dates as in the context of disabled input or options list.

**Comment by bdryanovski** [comment]
My primary request for changes is for the `aria-live` everything else is optional :)

File: src/dev/src/app/datepicker/datepicker-min-max.ts [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+ */
+import { Component } from '@angular/core';
+
+@Component({
+  selector: 'clr-datepicker-min-max-demo',
+  templateUrl: './datepicker-min-max.html',
+  styleUrls: ['./datepicker.demo.scss'],
+})
+export class DatepickerMinMaxDemo {
+  minDate: string = '2019-11-17';
```
[the remark this answers is not in the document]
Comment (hippee-lee) [reply src/dev/src/app/datepicker/datepicker-min-max.ts]: Is this something that should block this? I think that demonstrating how to use the date range is the important thing which this does.  I would rather have a `dumb` demo that shows how to use the feature.

[the remark this answers is not in the document]
Comment (bdryanovski) [reply src/dev/src/app/datepicker/datepicker-min-max.ts]: no, it's not a blocker it's more on the example/documentation part

File: src/clr-angular/forms/datepicker/providers/date-io.service.ts [hunk 3]
```diff
@@ -33,6 +42,30 @@ export class DateIOService {
...
   }
 
+  public setMinDate(date: string): void {
+    // NOTE: I'm expecting consumers to pass one of four things here:
+    //       A proper date string(2019-11-11), null, undefined or empty string ('')
+    if (!date) {
+      // attribute binding was removed, reset back to the beginning of time
+      this.disabledDates.minDate = new DayModel(0, 0, 1);
+    } else {
+      const [year, month, day] = date.split('-').map(n => parseInt(n, 10));
```
[the remark this answers is not in the document]
Comment (hippee-lee) [reply src/clr-angular/forms/datepicker/providers/date-io.service.ts]: The input format is for a format that is standard html spec. I discussed this a with with @gnomeontherun and we decided not to add this at this time.

File: src/dev/src/app/datepicker/datepicker-min-max.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (hippee-lee) [reply src/dev/src/app/datepicker/datepicker-min-max.ts]: ok, i would prefer to keep it simple for now. imo - having this extra functionality just to have a current date isn't needed here.

File: src/clr-angular/forms/datepicker/yearpicker.ts [hunk 4]
```diff
@@ -19,6 +19,9 @@ import { AriaLiveService } from '../../utils/a11y/aria-live.service';
   providers: [AriaLiveService],
   template: `
         <div class="year-switchers">
+          <div aria-live="polite" class="clr-sr-only">
```
[the remark this answers is not in the document]
Comment (hippee-lee) [reply src/clr-angular/forms/datepicker/yearpicker.ts]: I'm not seeing this in my current branch. I'll double check and make sure the latest changes are pushed.

File: src/clr-angular/forms/datepicker/yearpicker.spec.ts [hunk 5]
```diff
@@ -148,7 +147,7 @@ export default function() {
       it('calls to navigate to the previous decade', () => {
         spyOn(context.clarityDirective, 'previousDecade');
         const switchers: HTMLElement = context.clarityElement.querySelector('.year-switchers');
-        const button: HTMLButtonElement = <HTMLButtonElement>switchers.children[0];
+        const button: HTMLButtonElement = <HTMLButtonElement>switchers.children[1];
```
[the remark this answers is not in the document]
Comment (hippee-lee) [reply src/clr-angular/forms/datepicker/yearpicker.spec.ts]: I'm not seeing this in my current branch. I'll double check and make sure the latest changes are pushed.

File: src/clr-angular/forms/datepicker/monthpicker.ts [hunk 6]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2016-2018 VMware, Inc. All Rights Reserved.
+ * Copyright (c) 2016-2019 VMware, Inc. All Rights Reserved.
```
[the remark this answers is not in the document]
Comment (hippee-lee) [reply src/clr-angular/forms/datepicker/monthpicker.ts]: I'll revert the file.

File: src/clr-angular/forms/datepicker/yearpicker.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (bdryanovski) [reply src/clr-angular/forms/datepicker/yearpicker.ts]: You have this at the bottom of the file
```ts
// ...
  private updateRange(yrm: YearRangeModel): void {
    const floor = yrm.yearRange[0];
    const ceil = yrm.yearRange[yrm.yearRange.length - 1];
    this.decadeRange = `${floor} to ${ceil}`;
    this.ariaLiveService.announce(this.ariaLiveDecadeText);
  }
// ...
ngAfterViewInit() {
    this._datepickerFocusService.focusCell(this._elRef);
    this.ariaLiveService.announce(this.ariaLiveDecadeText);
  }
```
That will handle the `aria-live=polite`

File: src/clr-angular/forms/datepicker/yearpicker.spec.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (bdryanovski) [reply src/clr-angular/forms/datepicker/yearpicker.spec.ts]: The tests were failing cause there is no longer `aria-live` div element inside the template. And the tests are depending on the order and count of divs. But if we remove the `div` they will pass with there original values starting from `0`

File: src/dev/src/app/datepicker/datepicker-min-max.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (bdryanovski) [reply src/dev/src/app/datepicker/datepicker-min-max.ts]: I'm fine with that :)

File: src/clr-angular/forms/datepicker/yearpicker.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (hippee-lee) [reply src/clr-angular/forms/datepicker/yearpicker.ts]: k, something seems off. The working branch that I pushed seems different from this one. What I meant is that this line has already been removed in my working branch so I'm not sure why it is still here.

[the remark this answers is not in the document]
Comment (hippee-lee) [reply src/clr-angular/forms/datepicker/yearpicker.ts]: I found out I was looking at the wrong branch. I'll remove this and update the specs for the button order in the DOM.

File: src/clr-angular/forms/datepicker/yearpicker.spec.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (hippee-lee) [reply src/clr-angular/forms/datepicker/yearpicker.spec.ts]: I found out I was looking at the wrong branch. I'll remove this and update the specs for the button order in the DOM.

File: src/clr-angular/forms/datepicker/model/calendar-view.model.ts (hunk 1 above)
Comment (hippee-lee) [reply src/clr-angular/forms/datepicker/model/calendar-view.model.ts]: I'll update them for excluded.

File: src/clr-angular/forms/datepicker/date-input.spec.ts [hunk 7]
```diff
@@ -151,6 +151,22 @@ export default function() {
...
+          expect(testMinDateModel).toEqual(dateIOService.disabledDates.minDate);
+        });
+
+        it('should handle big endien date strings for max inputs', () => {
+          // NOTE: big endian date format is yyyy-mm-dd
+          const testComponent = context.fixture.componentInstance;
+          const [maxYear, maxMonth, maxDay] = testComponent.maxDate.split('-').map(n => parseInt(n, 10));
+          const testMaxDateModel = new DayModel(maxYear, maxMonth - 1, maxDay);
+          expect(testMaxDateModel).toEqual(dateIOService.disabledDates.maxDate);
+        });
```
Comment (gnomeontherun) [inline src/clr-angular/forms/datepicker/date-input.spec.ts]: Small ask, can we add a test to check that it handles invalid input?

Comment (gnomeontherun) [reply src/clr-angular/forms/datepicker/date-input.spec.ts]: Never mind, it is handled below.

**Pull request merged**